doc/obrms.1
.Dd May 26, 2026
.Dt OBRMS 1 URM
.Os "Open Babel" 3.2
.Sh NAME
.Nm obrms
.Nd "Calculate the heavy-atom RMSD between two chemically identical structures"
.Sh SYNOPSIS
.Nm
.Op -fmxso
.Ar filename1
.Ar filename2
.Sh DESCRIPTION
Computes the heavy\-atom RMSD of identical compound structures.
Structures in multi\-structure files are compared one\-by\-one unless
.Fl -firstonly
is passed, in which case only the first structure in the reference file is used.
.Sh OPTIONS
.Bl -tag -width flag
.It Fl f , -firstonly
use only the first structure in the reference file (otherwise the n\-th molecule
in the reference file is compared to the n\-th molecule in the test file)
.It Fl m ,  -minimize
Compute the minimum RMSD achievable by applying a rigid\-body transformation to
the test molecule
.It Fl x , -cross
Compute all n\(ha2 RMSDs between molecules of the reference file.
Test file is ignored.
.It Fl s , -separate
Separate reference file into constituent molecules (disconnected fragments)
and report best RMSD
.It Fl t Ar t , Fl -timeout Ar t
Give up on matching after specified number of seconds
.It Fl o Ar file , Fl -out Ar file
Re\-oriented test structure output (used with
.Fl m )
.El
.Sh EXAMPLES
.Dl "obrms ref.sdf test.sdf"
.Pp
Calculate the RMSD between ref.sdf and test.sdf.
Both files should have the same number of molecules.
.Dl "obrms -f ref.sdf test.sdf"
.Pp
Calculate the RMSD between the first molecule in ref.sdf and each molecule in
test.sdf.
.Sh SEE ALSO
.Xr obfit 1 .
.Pp
The web pages for Open Babel can be found at:
\%<\fBhttps://openbabel.org/\fR>
.Sh AUTHORS
.An -nosplit
Open Babel is developed by a cast of many, including current maintainers
.An Geoff Hutchison ,
.An Chris Morley ,
.An Michael Banck ,
and innumerable others who have contributed fixes and additions.
For more contributors to Open Babel, see
\%<\fBhttps://openbabel.org/docs/Introduction/thanks.html\fR>
.Sh COPYRIGHT
Copyright (C) 1998\-2001 by OpenEye Scientific Software, Inc.
.br
Some portions Copyright (C) 2001\-2007 by Geoffrey R. Hutchison and
other contributors.
.Pp
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation version 2 of the License.
.Pp
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.


NEWS
Open Babel 2.5.0 (2017-XX-XX)

Open Babel 2.4.0 (2016-9-15)

Open Babel 2.3.0 (2010-10-23)

This release represents a major update and should be a stable upgrade,
strongly recommended for all users of Open Babel. Highlights include a
completely rewritten stereochemistry engine, Spectrophore fingerprint
generation, 2D depiction, improved 3D coordinate generation, conformer
searching, and more. Many formats are improved or added, including
CIF, PDBQT, SVG, and more. Improved developer API and scripting
support and many, many bug fixes are also included.

What's New from 2.2.3:

* Completely rewritten stereochemistry perception, including support
  for tetrahedral, square planar, and higher-order stereochemistry.
* Dramatically improved canonicalization algorithm (Note that in general, canonical SMILES have changed since the 2.2.x release.)
* 2D depiction, including SVG vector graphics generation using code from MCDL.
* New Spectrophore generation, contributed by Silicos NV.
* New ChargeMethod API including support for partial charge assignment
  from Gasteiger, MMFF94, QEq, QTPIE methods and plugin interface for
  adding more.
* Improved 3D coordinate generation.
* New conformer generation framework, including support for diverse
  conformer generation and genetic algorithm lowest-energy searching.
* Improved user documentation.
* Improved aromaticity / Kekule bond assignment.
* Improved unit test suite using the CMake-based CTest program.
* Improved support for crystallographic unit cells (e.g., in CIF format).
* Improved UFF force field method, including hypervalent 5, 6, 7 and higher coordination numbers.
* Support for the GAFF (Generalized Amber Force Field) method.
* Support for reading geometry optimizations as multiple conformers
  from Gaussian, GAMESS-US, and other quantum chemistry packages.
* Support for reading molecular orbital energies from quantum chemistry formats.
* Several memory leaks fixed.
* Fixed many compiler warnings.
* Fixed support for MinGW and Cygwin environments.
* Fixed bugs with Gaussian 09 output files.
* Latest released version of the InChI library (1.0.3) generating standard InChI.
* Many more bug fixes and small feature improvements.

==New Command-Line Operations==

* --canonical: Output atoms in canonical order for any file format (i.e., not just SMILES)
* --conformer: Run a conformer search on the input molecules (has many options)
* --gen2D: Generate a 2D depiction of the molecule
* --partialcharge <model>: Use the partial charge model supplied to
  generate charges (i.e., instead of default Gasteiger sigma model)
* --sort <descriptor>: Sort molecules by a specified descriptor
* --unique: Only output unique molecules (as determined by InChI generation)

==New File Formats==

Import & Export:
- DL-POLY CONFIG
- FHIaims XYZ
- PDBQT

Import only:
- DL-POLY HISTORY
- GULP output
- PWscf output
- Text

Export only:
- MNA (Multilevel Neighborhoods of Atoms)
- SVG vector graphics


Open Babel 2.2.3 (2009-07-31)

What's New from 2.2.2:

This release represents an important bug-fix upgrade, strongly recommended
for all users of Open Babel.

- Fixed bug in fingerprints in 2.2.2, where the default fingerprint
  format and bit length was changed inadvertently.
- Fixed detection of shared_ptr in tr1/memory.
- Fixed additional aromaticity / Kekule assignment bugs.
- Fixed several bugs in the MMCIF format.
- Additional bug fixes.

Open Babel 2.2.2 (2009-07-04)

What's New from 2.2.1:

This release represents a major bug-fix release and is a stable
upgrade, strongly recommended for all users of Open Babel. While there
may not be many new features, many crashes and other bugs have been
fixed since 2.2.1.

- Upgraded to the new InChI 1.02 release to produce standardized InChI
  and InChIKey output.
- Fixed many stereochemistry errors when reading/writing SMILES. This
  is part of a larger project which will be finished in the 2.3
  release.
- Fixed compilation and installation on Cygwin and MinGW platforms.
- Significantly improved aromaticity and Kekule bond assignment.
- Improved 2D -> 3D coordinate generation
- Improved coordinate generation using the --gen3d command-line
  operation
- Improved performance for coordinate generation.
- New --fillUC command-line operation for babel.
- Fixes to pH-dependent hydrogen addition.
- Added support for reading vibrational data from Molden, Molpro, and
  NWChem output files.
- Updated atomic radii from recent theoretical calculations.
- Fixed bug when reading gzip-compressed Mol2 or XML files.
- Close files after an error. Fixes a bug with Pybel where files would
  remain open.
- Many more bug fixes and small feature improvements.

==New File Formats==

Import & Export:
- Molpro input and output.
- VASP coordinate files (CONTCAR and POSCAR).

Open Babel 2.2.1 (2009-03-01)

What's New from 2.2.0:

This release represents a major bug-fix release and is a stable
upgrade, strongly recommended for all users of Open Babel. While there
may not be many new features, many crashes and other bugs have been
fixed since 2.2.0.

- Improved scripting interfaces, including Python 3 support and improved
  Java and C# support.
- Added support for MACCS fingerprints. Thanks to the RDKit project.
- Many fixes and enhancements to the force field code. In particular,
  the UFF force field implementation should handle many more molecules.
- Improved 3D coordinate generation, particularly with ring
  fragments. You can give this a try with the obgen utility.
- Fixed a variety of PDB import errors with atom types.
- Added support for reading charges and radii from PQR file formats.
- Added support for reading and writing unit cells in PDB formats.
- New "output" file format for taking generic ".out", ".log", and
  ".dat" files and reading with appropriate file type based on
  contents. Currently works extremely well for quantum chemistry
  packages.
- Added improved error handling and reporting when unable to load file
  formats.

tools/obrms.cpp
/**********************************************************************
 obrms = Returns the rms between two chemically identical structures
 Derived from obfit.
 *
 This file is part of the Open Babel project.
 For more information, see <http://openbabel.org/>
 
 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation version 2 of the License.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 ***********************************************************************/

/*
 Require a fixed molecule and a set of molecules to compare against.
 example of command line:
 obrms ref.sdf test.sdf
 */

#undef _GLIBCXX_DEBUG /* unless you have boost libraries built with this, you do _not_ want this defined */
// used to set import/export for Cygwin DLLs
#ifdef WIN32
#define USING_OBDLL
#endif
#include <cstdlib>
#include <memory>
#include <openbabel/babelconfig.h>
#include <openbabel/mol.h>
#include <openbabel/parsmart.h>
#include <openbabel/obconversion.h>
#include <openbabel/query.h>
#include <openbabel/isomorphism.h>
#include <openbabel/obutil.h>

#include "getopt.h"

#include <sstream>

using namespace std;
using namespace OpenBabel;

class AtomDistanceSorter
{
	vector3 ref;
	public:
	AtomDistanceSorter(OBAtom *r) :
			ref(r->GetVector())
	{

	}
	bool operator()(OBAtom *l, OBAtom *r) const
			{
		double ld = ref.distSq(l->GetVector());
		double rd = ref.distSq(r->GetVector());

		return ld < rd;
	}
};
/* class to perform graph matching between two molecules.
 * Is initialized with the reference molecule.
 * Will figure out the atom correspondences and compute the rmsd between the
 * ref mol and a provided test mol.
 */
class Matcher
{
	const OBMol *ref;
	std::shared_ptr<OBQuery> query;
	std::shared_ptr<OBIsomorphismMapper> mapper;

	class MapRMSDFunctor: public OBIsomorphismMapper::Functor
	{
	private:
		const OBMol *ref;
		OBMol& test;
		double bestRMSD;
		bool minimize;
		public:
		//will modify t if min is true
		MapRMSDFunctor(const OBMol* r, OBMol& t, bool min = false) :
				ref(r), test(t), bestRMSD(HUGE_VAL), minimize(min)
		{
		}

		bool operator()(OBIsomorphismMapper::Mapping &map) override
		{
			unsigned N = map.size();
			double *refcoord = (double*)alloca(sizeof(double)*N * 3);
			double *testcoord = (double*)alloca(sizeof(double)*N * 3);

			for (unsigned i = 0; i < N; i++)
			{
				//obmol indices are 1-indexed while the mapper is zero indexed
				const OBAtom *ratom = ref->GetAtom(map[i].first + 1);
				const OBAtom *tatom = test.GetAtom(map[i].second + 1);
				assert(ratom && tatom);

				for (unsigned c = 0; c < 3; c++)
				{
					refcoord[3 * i + c] = ratom->GetVector()[c];
					testcoord[3 * i + c] = tatom->GetVector()[c];
				}
			}

			if (minimize)
			{
				double rmatrix[3][3] =
						{ 0 };

				double rave[3] =
						{ 0, 0, 0 };
				double tave[3] =
						{ 0, 0, 0 };
				//center
				for (unsigned i = 0; i < N; i++)
				{
					for (unsigned c = 0; c < 3; c++)
					{
						rave[c] += refcoord[3 * i + c];
						tave[c] += testcoord[3 * i + c];
					}
				}

				for (unsigned c = 0; c < 3; c++)
				{
					rave[c] /= N;
					tave[c] /= N;
				}

				for (unsigned i = 0; i < N; i++)
				{
					for (unsigned c = 0; c < 3; c++)
					{
						refcoord[3 * i + c] -= rave[c];
						testcoord[3 * i + c] -= tave[c];
					}
				}

				qtrfit(refcoord, testcoord, N, rmatrix);
				rotate_coords(testcoord, rmatrix, N); 

				for (unsigned i = 0; i < N; i++)
				{
					//with minimize on, change coordinates
					OBAtom *tatom = test.GetAtom(map[i].second + 1);
					tatom->SetVector(testcoord[3*i]+rave[0], testcoord[3*i+1]+rave[1], testcoord[3*i+2]+rave[2]);
				}
			}

			double rmsd = calc_rms(refcoord, testcoord, N);

			if (rmsd < bestRMSD)
				bestRMSD = rmsd;
			// check all possible mappings
			return false;
		}

		double getRMSD() const
		{
			return bestRMSD;
		}
	};

public:
	Matcher(OBMol& mol, int timeout) : ref(&mol)
	{
		query = std::shared_ptr<OBQuery>(CompileMoleculeQuery(&mol));
		mapper = std::shared_ptr<OBIsomorphismMapper>(OBIsomorphismMapper::GetInstance(query.get()));
		if(timeout > 0) {
			mapper->SetTimeout(timeout);
		}
	}


	//computes a correspondence between the ref mol and test (exhaustively)
	//and returns the rmsd; returns infinity if unmatchable
	double computeRMSD(OBMol& test, bool minimize = false)
	{
		MapRMSDFunctor funct(ref, test, minimize);

		mapper->MapGeneric(funct, &test);
		return funct.getRMSD();
	}
};

//preprocess molecule into a standardized state for heavy atom rmsd computation
static void processMol(OBMol& mol)
{
	//isomorphismmapper wants isomorphic atoms to have the same aromatic and ring state,
	//but these proporties aren't reliable enough to be trusted in evaluating molecules
	//should be considered the same based solely on connectivity

	mol.DeleteHydrogens(); //heavy atom rmsd
	for (OBAtomIterator aitr = mol.BeginAtoms(); aitr != mol.EndAtoms(); aitr++)
	{
		OBAtom *a = *aitr;
		a->SetAromatic(false);
		a->SetInRing();
	}
	for (OBBondIterator bitr = mol.BeginBonds(); bitr != mol.EndBonds(); bitr++)
	{
		OBBond *b = *bitr;
		b->SetAromatic(false);
		b->SetBondOrder(1);
		b->SetInRing();
	}
	//avoid recomputations
	mol.SetHybridizationPerceived();
	mol.SetRingAtomsAndBondsPerceived();
	mol.SetAromaticPerceived();
}
///////////////////////////////////////////////////////////////////////////////
//! \brief compute rms between chemically identical molecules
int main(int argc, char **argv)
{
	bool firstOnly = false;
	bool minimize = false;
	bool separate = false;
	bool help = false;
	bool docross = false;
	string fileRef;
	string fileTest;
	string fileOut;
	int timeout = 0;

	const char *helpmsg =
	 "obrms: Computes the heavy-atom RMSD of identical compound structures.\n"
	  "Usage: obrms reference_file [test_file]\n"
	  "Options:\n"
	  "\t -o file, --out  file       re-oriented test structure output\n"
	  "\t -f,      --firstonly       use only the first structure in the reference file\n"
	  "\t -m,      --minimize        compute minimum RMSD\n"
	  "\t -x,      --cross           compute all n^2 RMSDs between molecules of reference file\n"
	  "\t -s,      --separate        separate reference file into constituent molecules and report best RMSD\n"
	  "\t -t t,    --timeout t       give up on matching after specified number of seconds\n"
	  "\t -h,      --help            help message\n";
	struct option long_options[] = {
	    {"firstonly", no_argument, nullptr, 'f'},
	    {"minimize", no_argument, nullptr, 'm'},
	    {"cross", no_argument, nullptr, 'x'},
	    {"separate", no_argument, nullptr, 's'},
	    {"timeout", required_argument, nullptr, 't'},
	    {"out", required_argument, nullptr, 'o'},
	    {"help", no_argument, nullptr, 'h'},
	    {nullptr, 0, nullptr, 0}
	};
	int option_index = 0;
	int c = 0;
	while ((c = getopt_long(argc, argv, "hfmxst:o:", long_options, &option_index) ) != -1) {
	  switch(c) {
	    case 'o':
	      fileOut = optarg;
	      break;
	    case 'f':
	      firstOnly = true;
	      break;
	    case 'm':
	      minimize = true;
	      break;
	    case 'x':
	      docross = true;
	      break;
	    case 's':
	      separate = true;
	      break;
		case 't':
		  timeout = atoi(optarg);
		  break;
	    case 'h':
	      cout << helpmsg;
	      exit(0);
	      break;
	    default:
	      cerr << "Unrecognized option: " << c << "\n";
	      exit(-1);
	  }
	}

	if(optind < argc) {
	  fileRef = argv[optind];
	  optind++;
	}
	if(optind < argc) {
	  fileTest = argv[optind];
	  optind++;
	}

	if(optind < argc) {
	  cerr << "Unrecognized argument: " << argv[optind];
	  exit(-1);
	}

	if(!docross && fileTest.size() == 0) {
	  cerr << helpmsg;
	  cerr << "Command line parse error: test file is required but missing\n";
	  exit(-1);
	}

	//open mols
	OBConversion refconv(fileRef);

	OBConversion outconv;
	OBFormat *outFormat = outconv.FormatFromExt(fileOut);
	if(fileOut.size() > 0)
	{
		if(!outFormat || !outconv.SetInAndOutFormats(outFormat, outFormat))
		{
			cerr << "Do not understand output format!" << endl;
			exit(-1);
		}
	}

	std::ofstream out;
	if(fileOut.size() > 0)
	{
		out.open(fileOut.c_str());
	}


  //read reference
  OBMol molref;

	if(docross) {
	  //load in the entire reference file
    vector<OBMol> refmols;
    while (refconv.Read(&molref))
    {
       processMol(molref);
       refmols.push_back(molref);
    }

    for(unsigned i = 0, n = refmols.size() ; i < n; i++) {
      OBMol& ref = refmols[i];
      Matcher matcher(ref, timeout);
      cout << ref.GetTitle();
      for(unsigned j = 0; j < n; j++) {
        OBMol& moltest = refmols[j];
        double rmsd = matcher.computeRMSD(moltest, minimize);
        cout << ", " << rmsd;
      }
      cout << "\n";
    }

	} else {

    OBConversion testconv(fileTest);

	  //check comparison file
    while (refconv.Read(&molref))
    {
      vector<OBMol> refmols;
      if(separate) {
        refmols = molref.Separate();
      } else {
        refmols.push_back(molref);
      }

      vector<Matcher> matchers;
      for(unsigned i = 0, n = refmols.size(); i < n; i++) {
        processMol(refmols[i]);
        Matcher matcher(refmols[i], timeout); // create the matcher
        matchers.push_back(matcher);
      }

      OBMol moltest;
      while (testconv.Read(&moltest))
      {
        if (moltest.Empty())
          break;

        processMol(moltest);

        double bestRMSD = HUGE_VAL;
        for(unsigned i = 0, n = matchers.size(); i < n; i++) {
          double rmsd = matchers[i].computeRMSD(moltest, minimize);
          if(rmsd < bestRMSD) bestRMSD = rmsd;
        }

        cout << "RMSD " << molref.GetTitle() << ":" <<  moltest.GetTitle() << " " << bestRMSD << "\n";

        if(out)
        {
          outconv.Write(&moltest, &out);
        }
        if (!firstOnly) //one test molecule will be read for each reference molecule
          break;
      }
      if(firstOnly) //done with first reference mol
        break;
    }
	}
	return (0);
}
